lib/r10k/api.rb

```diff
@@ -186,22 +186,66 @@ def module_source_for_module(module_name, env_map)
     def module_sources_for_environment(env_map)
     end
 
+    # Update local caches represented by the given by module_sources, a collection of module_source hashmaps.
+    #
+    # @param module_sources [Array<Hash>] An array of hashmaps each representing a single remote module source (as produced by {#module_sources_for_environment})
+    # @param base_cachedir [String] Root of where r10k is caching things.
+    # @return [true] Returns true on success, raises on failure.
+    # @raise [RuntimeError] Something bad happened!
+    def update_caches(module_sources, base_cachedir, opts = {})
+      if module_sources.respond_to?(:each)
+        module_sources.each do |module_source|
+          update_cache(module_source, base_cachedir, opts)
+        end
+        return true
+      else
+        raise RuntimeError.new("module_sources must be a collection of source hashes.")
+      end
+
+    end
     # Update local cache represented by the given module_source hashmap.
     #
     # @param module_source [Hash] A hashmap representing a single remote module source (as produced by {#module_sources_for_environment})
+    # @param base_cachedir [String] Root of where r10k is caching things.
     # @return [true] Returns true on success, raises on failure.
-    # @raise [RuntimeError] Something bad happened!
-    def update_cache(module_source)
+    # @raise [UnresolvableError] Something bad happened!
+    def update_cache(module_source, base_cachedir, opts = {})
+      type = module_source[:type]
+      source = module_source[:source]
+
+      raise UnresolvableError.new('module source type cannot be nil.') if type.nil?
```

Comment:
I think UnresolvableError should probably be reserved for the method(s) that specifically attempt envmap resolution